File: scripts/linters/no_use_as_alias.rb
def no_use_as_alias(root_dir, excludes = [])
  pattern = root_dir.join('crates', '**', '*.rs').to_s
  errors = Dir.glob(pattern).sort.flat_map do |path|
    relative = Pathname.new(path).relative_path_from(root_dir).to_s
    next [] if excludes.include?(relative)

    find_use_aliases(path, relative)
  end

  return true if errors.empty?

  puts 'Found `use ... as name` aliases.'
  puts 'Aliasing imports merely to shorten a namespace is forbidden.'
  puts 'Only `as _` (e.g. `use std::io::Write as _;`) and PascalCase renames'
  puts '(for collision avoidance, e.g. `use foo::Error as FooError;`) are allowed:'
  errors.each do |err|
    puts "  #{err}"
  end
  false
end

USE_ALIAS_START_RE = /\A(?:pub(?:\([^)]*\))?\s+)?use\b/
PASCAL_CASE_RE = /\A[A-Z][A-Za-z0-9]*\z/

def find_use_aliases(path, relative)
  errors = []
  in_use = false
  brace_depth = 0

  File.readlines(path).each_with_index do |raw, idx|
    code = raw.split('//', 2).first || raw
    stripped = code.strip

    unless in_use
      next unless stripped =~ USE_ALIAS_START_RE

      in_use = true
      brace_depth = 0
    end

    code.scan(/\bas\s+([A-Za-z_][A-Za-z0-9_]*)/) do |m|
      name = m[0]
      next if name == '_'
      next if name =~ PASCAL_CASE_RE

      errors << "#{relative}:#{idx + 1}: `as #{name}` aliasing in `use` statement"
    end

    brace_depth += code.count('{') - code.count('}')
    if brace_depth <= 0 && code.rstrip.end_with?(';')
      in_use = false
      brace_depth = 0
    end
  end

  errors
end

File: scripts/linters/src/Linters/NoUseAsAlias.php
<?php

declare(strict_types=1);

namespace Shirabe\Lint\Linters;

use Shirabe\Lint\Linter;
use Shirabe\Lint\Support\FileFinder;
use Shirabe\Lint\Support\Paths;

final class NoUseAsAlias implements Linter
{
    private const USE_ALIAS_START_RE = '/\A(?:pub(?:\([^)]*\))?\s+)?use\b/';
    private const PASCAL_CASE_RE = '/\A[A-Z][A-Za-z0-9]*\z/';

    public function name(): string
    {
        return 'no_use_as_alias';
    }

    public function failureIntro(): string
    {
        return "Found `use ... as name` aliases.\n"
            . "Aliasing imports merely to shorten a namespace is forbidden.\n"
            . "Only `as _` (e.g. `use std::io::Write as _;`) and PascalCase renames\n"
            . '(for collision avoidance, e.g. `use foo::Error as FooError;`) are allowed:';
    }

    public function check(string $rootDir, array $excludes): array
    {
        $errors = [];

        foreach (FileFinder::rustFiles($rootDir) as $path) {
            $relative = Paths::relativeTo($rootDir, $path);
            if (in_array($relative, $excludes, true)) {
                continue;
            }

            array_push($errors, ...$this->findUseAliases($path, $relative));
        }

        return $errors;
    }

    /** @return list<string> */
    private function findUseAliases(string $path, string $relative): array
    {
        $errors = [];
        $inUse = false;
        $braceDepth = 0;

        foreach (file($path) as $idx => $raw) {
            $code = explode('//', $raw, 2)[0];
            $stripped = trim($code);

            if (!$inUse) {
                if (!preg_match(self::USE_ALIAS_START_RE, $stripped)) {
                    continue;
                }
                $inUse = true;
                $braceDepth = 0;
            }

            if (preg_match_all('/\bas\s+([A-Za-z_][A-Za-z0-9_]*)/', $code, $m)) {
                foreach ($m[1] as $name) {
                    if ($name === '_') {
                        continue;
                    }
                    if (preg_match(self::PASCAL_CASE_RE, $name)) {
                        continue;
                    }
                    $errors[] = "{$relative}:" . ($idx + 1) . ": `as {$name}` aliasing in `use` statement";
                }
            }

            $braceDepth += substr_count($code, '{') - substr_count($code, '}');
            if ($braceDepth <= 0 && str_ends_with(rtrim($code), ';')) {
                $inUse = false;
                $braceDepth = 0;
            }
        }

        return $errors;
    }
}
